### PR TITLE
A few extra step function utility methods.

### PR DESCRIPTION
**Pull request by xbrianh**
Utility methods needed for visitation batch processor, may
be useful for other step function consumers.

connects to #539

### Test plan
Tested locally through visitation batch processor integration tests. Will be tested through smoketest in the future.

### CODE REVIEW

File: dss/stepfunctions/__init__.py [hunk 1]
```diff
@@ -16,18 +54,84 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
     )
     return response
+
+
+def step_functions_describe_execution(state_machine_name_template: str, execution_name: str) -> typing.Any:
+    """
+    Return description of a step function exectuion, possible in-progress, completed, errored, etc.
+    :param state_machine_name_template:
+    :param execution_name:
+    :return:
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: Is this redundant?

File: dss/stepfunctions/__init__.py [hunk 2]
```diff
@@ -16,18 +54,84 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
+        raise Exception('Need at least 1 AWS api call')
+
+    executions = list()
+
+    kwargs = {
+        'stateMachineArn': state_machine_arn,
+        'maxResults': k_results_per_page
+    }
+
+    for k_api_calls in range(1, k_max_pages + 1):
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: `range(k_max_pages)` does the same, correct?

Comment (hannes-ucsc) [reply dss/stepfunctions/__init__.py]: Also, are you 1-based?

;-)

File: dss/stepfunctions/__init__.py [hunk 3]
```diff
@@ -16,18 +54,84 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
+            if execs[-1]['startDate'] < start_date:
+                executions.extend([e for e in execs
+                                   if e['startDate'] > start_date])
+                break
+            else:
+                executions.extend(execs)
+
+        if not resp.get('nextToken', False):
+            break
+
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: I would add `else:` here to be consistent and explicit.

And then I would remove `not` and swap the branches.

File: dss/stepfunctions/__init__.py [hunk 4]
```diff
@@ -16,18 +54,84 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
+        'maxResults': k_results_per_page
+    }
+
+    for k_api_calls in range(1, k_max_pages + 1):
+        resp = sfn.list_executions(**kwargs)
+
+        execs = resp['executions']
+
+        if execs:
+            if execs[-1]['startDate'] < start_date:
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: This optimization depends on the fact that the result is sorted. Is that documented or incidental?

Comment (hannes-ucsc) [reply dss/stepfunctions/__init__.py]: In fact, for robustness, I would not optimize and take the performance hit.

File: dss/stepfunctions/__init__.py (hunk 2 above)
Comment (xbrianh) [reply dss/stepfunctions/__init__.py]: Heh... I was returning k_api_calls previously, thus wanted at least 1.

File: dss/stepfunctions/__init__.py (hunk 4 above)
Comment (xbrianh) [reply dss/stepfunctions/__init__.py]: Good eyes, Hannes.

File: dss/stepfunctions/__init__.py [hunk 5]
```diff
@@ -1,10 +1,48 @@
...
+    accountid = sts_client.get_caller_identity()['Account']
+    stage = os.environ["DSS_DEPLOYMENT_STAGE"]
+    sfn_name = state_machine_name_template.format(stage=stage)
+    state_machine_arn = f"arn:aws:states:{region}:{accountid}:stateMachine:{sfn_name}"
+    return state_machine_arn
+
+
+def step_functions_execution_arn(state_machine_name_template: str, execution_name: str) -> str:
+    """
+    The ARN of a state machine executiuon, with name derived from `state_machine_name_template`, with string formatting
```
Comment (ttung) [inline dss/stepfunctions/__init__.py]: *execution

File: dss/stepfunctions/__init__.py [hunk 6]
```diff
@@ -16,18 +54,84 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
+
+    resp = sfn.describe_execution(
+        executionArn=execution_arn
+    )
+
+    return resp
+
+
+def step_functions_list_executions(state_machine_name_template: str,
+                                   start_date: typing.Type(datetime.datetime)=None,
```
Comment (ttung) [inline dss/stepfunctions/__init__.py]: you can just do:

`start_date: datetime.datetime=None,`

File: dss/stepfunctions/__init__.py [hunk 7]
```diff
@@ -16,18 +54,84 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
         stateMachineArn=state_machine_arn,
         name=execution_name,
         input=execution_input
     )
     return response
+
+
+def step_functions_describe_execution(state_machine_name_template: str, execution_name: str) -> typing.Any:
+    """
+    Return description of a step function exectuion, possible in-progress, completed, errored, etc.
```
Comment (ttung) [inline dss/stepfunctions/__init__.py]: *execution

**Comment by ttung** [comment]
Also, what's the state of this PR?

**Comment by xbrianh** [comment]
Re-opened with date filtering removed from `list_executions`, using a generator pattern with backgrounded paging instead.

File: dss/stepfunctions/__init__.py [hunk 8]
```diff
@@ -1,10 +1,48 @@
...
 
 import boto3
 
 
+def step_functions_arn(state_machine_name_template: str) -> str:
+    """
+    The ARN of a state machine, with name derived from `state_machine_name_template`, with string formatting to
+    replace {stage} with the dss deployment stage.
+    :param state_machine_name_template:
+    :return:
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: remove this line?

File: dss/stepfunctions/__init__.py [hunk 9]
```diff
@@ -1,10 +1,48 @@
...
+    return state_machine_arn
+
+
+def step_functions_execution_arn(state_machine_name_template: str, execution_name: str) -> str:
+    """
+    The ARN of a state machine execution, with name derived from `state_machine_name_template`, with string formatting
+    to replace {stage} with the dss deployment stage.
+    :param state_machine_name_template:
+    :param execution_name:
+    :return:
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: Remove these 3 lines?

File: dss/stepfunctions/__init__.py (hunk 7 above)
Comment (hannes-ucsc) [reply dss/stepfunctions/__init__.py]: `possibly` ?

File: dss/stepfunctions/__init__.py (hunk 1 above)
Comment (hannes-ucsc) [reply dss/stepfunctions/__init__.py]: Remove these three lines?

File: dss/stepfunctions/__init__.py [hunk 10]
```diff
@@ -16,18 +54,62 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
+
+
+def step_functions_list_executions(state_machine_name_template: str, k_results_per_page: int=100) -> typing.Iterable:
+    """
+    List step function executions, peforming paging in the background.
+    Maximum 100 results per page.
+    :param state_machine_name_template:
+    :param k_results_per_page:
+    :return:
+    """
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: Remove these three lines?

File: dss/stepfunctions/__init__.py [hunk 11]
```diff
@@ -16,18 +54,62 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
+
+    sfn = boto3.client('stepfunctions')
+
+    state_machine_arn = step_functions_arn(state_machine_name_template)
+
+    kwargs = {
+        'stateMachineArn': state_machine_arn,
+        'maxResults': k_results_per_page
+    }
```
Comment (hannes-ucsc) [inline dss/stepfunctions/__init__.py]: Nit: I usually initialize kwargs using the `dict` constructor. The keys are bound to be valid Python identifiers and more closely resembles a direct invocation, without `**`.

File: dss/stepfunctions/__init__.py [hunk 12]
```diff
@@ -5,6 +5,43 @@
 import boto3
 
 
+def step_functions_arn(state_machine_name_template: str) -> str:
+    """
+    The ARN of a state machine, with name derived from `state_machine_name_template`, with string formatting to
+    replace {stage} with the dss deployment stage.
+    :param state_machine_name_template:
+    :return:
```
Comment (kislyuk) [inline dss/stepfunctions/__init__.py]: If you're adding signature docstrings, please fill them out. If your IDE has inserted a placeholder signature docstring but you don't want to fill it out, please delete it.

File: dss/stepfunctions/__init__.py [hunk 13]
```diff
@@ -5,6 +5,43 @@
...
+    """
+
+    sts_client = boto3.client("sts")
```
Comment (kislyuk) [inline dss/stepfunctions/__init__.py]: Here and elsewhere, please use `dss.util.aws.clients.sts` (and `dss.util.aws.resources`), which avoids re-initializing the client every time you run this function.

File: dss/stepfunctions/__init__.py [hunk 14]
```diff
@@ -5,6 +5,43 @@
...
+
+    region = boto3.Session().region_name
+    accountid = sts_client.get_caller_identity()['Account']
```
Comment (kislyuk) [inline dss/stepfunctions/__init__.py]: Here and elsewhere, please use `dss.util.aws.ARN.get_region()` and `dss.util.aws.ARN.get_account_id()`. This avoids an expensive network call every time you run this function. In general, please make sure you watch for unnecessary network calls.

File: dss/stepfunctions/__init__.py [hunk 15]
```diff
@@ -16,18 +53,62 @@ def step_functions_invoke(state_machine_name_template: str, execution_name: str,
...
+    while True:
+        resp = sfn.list_executions(**kwargs)
+
+        for e in resp['executions']:
+            yield e
+
+        if resp.get('nextToken', False):
+            kwargs['nextToken'] = resp['nextToken']
+        else:
+            break
```
Comment (kislyuk) [inline dss/stepfunctions/__init__.py]: Why is the pagination being done manually instead of using a boto3 paginator?

**Comment by hannes-ucsc** [comment]
Looks good to me, but there is this one line with a typo that @ttung pointed out.

File: dss/stepfunctions/__init__.py [hunk 16]
```diff
@@ -2,32 +2,95 @@
...
+
+
+def step_functions_describe_execution(state_machine_name_template: str, execution_name: str) -> typing.Any:
+    """
+    Return description of a step function execution, possible in-progress, completed, errored, etc.
+    """
+
+    execution_arn = step_functions_execution_arn(state_machine_name_template, execution_name)
+
+    resp = stepfunctions.describe_execution(
```
Comment (ttung) [inline dss/stepfunctions/__init__.py]: unwrap please.

File: dss/stepfunctions/__init__.py [hunk 17]
```diff
@@ -2,32 +2,95 @@
...
+
+
+def step_functions_list_executions(
+    state_machine_name_template: str,
+    status_filter: str=None,
+    k_results_per_page: int=None
+) -> typing.Iterable:
+    """
+    List step function executions, peforming paging in the background.
+    Maximum 100 results per page.
```
Comment (ttung) [inline dss/stepfunctions/__init__.py]: what's the significance of this comment ("Maximum 100 results per page")?

File: dss/stepfunctions/__init__.py [hunk 18]
```diff
@@ -2,32 +2,95 @@
...
+    """
+
+    state_machine_arn = step_functions_arn(state_machine_name_template)
+
+    kwargs = dict(
```
Comment (ttung) [inline dss/stepfunctions/__init__.py]: consider unwrapping.

File: dss/stepfunctions/__init__.py [hunk 19]
```diff
@@ -2,32 +2,87 @@
...
+    """
+    execution_arn = step_functions_execution_arn(state_machine_name_template, execution_name)
+    resp = stepfunctions.describe_execution(executionArn=execution_arn)
+    return resp
+
+
+def step_functions_list_executions(
+    state_machine_name_template: str,
+    status_filter: str=None,
+    k_results_per_page: int=None
```
Comment (kislyuk) [inline dss/stepfunctions/__init__.py]: Why is this prefixed with `k_`? Consider renaming it to `max_results_per_page`.

Comment (xbrianh) [reply dss/stepfunctions/__init__.py]: Because of a personal convention... renamed to `max_results_per_page`

**Comment by hannes-ucsc** [comment]
I'm gonna back this one out. It was merged with failing tests.

**Pull request merged**